ingress/controllers/nginx-alpha-ssl/controller.go

```diff
@@ -0,0 +1,217 @@
+
+{{range $ing := .Items}}
+{{range $rule := $ing.Spec.Rules}}
+	server {
+		server_name {{$rule.Host}};
+{{range $key, $val := $ing.Labels}}
+{{if (and (eq $key "ssl") (eq $val "true"))}}
+		listen 443 ssl;
+		ssl_certificate	/etc/nginx/certs/{{$rule.Host}}.crt;
+		ssl_certificate_key	/etc/nginx/certs/{{$rule.Host}}.key;
+{{- end}}{{end}}
+		error_log /dev/stdout info;
+		access_log /dev/stdout default;
+
+		listen 80;
+{{ range $path := $rule.HTTP.Paths }}
+		location {{$path.Path}} {
+			proxy_set_header Host $host;
+			proxy_pass http://{{$path.Backend.ServiceName}}.{{$ing.Namespace}}.svc.cluster.local:{{$path.Backend.ServicePort}};
+		}{{end}}
+	}{{end}}{{end}}
+}`
+)
+
+const nginxConfDir = "/etc/nginx"
+const nginxCommand = "nginx"
+
+// shellOut runs an external command.
+// stdout and stderr are attached to this external process
+func shellOut(shellCmd string, args []string) {
+	cmd := exec.Command(shellCmd, args...)
+	stdout, _ := cmd.StdoutPipe()
+	stderr, _ := cmd.StderrPipe()
+
+	fmt.Printf("Starting %v %v\n", shellCmd, args)
+
+	go io.Copy(os.Stdout, stdout)
+	go io.Copy(os.Stderr, stderr)
+	err := cmd.Start()
+	if err != nil {
+		log.Fatalf("Failed to execute %v: err: %v", cmd, err)
+	}
+}
+
+func main() {
+	var ingClient client.IngressInterface
+	if kubeClient, err := client.NewInCluster(); err != nil {
+		log.Fatalf("Failed to create client: %v.", err)
+	} else {
+		ingClient = kubeClient.Extensions().Ingress(api.NamespaceAll)
+	}
+	/* vaultEnabled
+	The following environment variables should be set:
+	VAULT_ADDR
+	VAULT_TOKEN
+	VAULT_SKIP_VERIFY (if using self-signed SSL on vault)
+	The only one we need to explicitly introduce is VAULT_ADDR, but we can check the others
+	*/
+	vaultEnabled := true
+	config := vault.DefaultConfig()
+	config.Address = os.Getenv("VAULT_ADDR")
+	vault, err := vault.NewClient(config)
+	if err != nil {
+		fmt.Printf("WARN: VAULT_ADDR is not set\n")
+		vaultEnabled = false
+	}
+
+	token := os.Getenv("VAULT_TOKEN")
+	if token == "" {
+		fmt.Printf("WARN: VAULT_TOKEN is not set\n")
+		vaultEnabled = false
+	}
+
+	tmpl, _ := template.New("nginx").Parse(nginxConf)
+	rateLimiter := util.NewTokenBucketRateLimiter(0.1, 1)
+	known := &extensions.IngressList{}
+
+	// Controller loop
+	nginxArgs := []string{
+		"-c",
+		nginxConfDir + "/nginx.conf",
+	}
+
+	shellOut(nginxCommand, nginxArgs)
+	for {
+		rateLimiter.Accept()
+		ingresses, err := ingClient.List(api.ListOptions{})
+		if err != nil {
+			fmt.Printf("Error retrieving ingresses: %v\n", err)
+			continue
+		}
+		if reflect.DeepEqual(ingresses.Items, known.Items) {
+			continue
+		}
+		known = ingresses
+
+		if vaultEnabled {
+			for _, ingress := range ingresses.Items {
+				ingressHost := ingress.Spec.Rules[0].Host
+				vaultPath := "secret/ssl/" + ingressHost
+
+				keySecretData, err := vault.Logical().Read(vaultPath)
```

Comment:
can you include some docs on how one inserts certs into value? for bonus points you can make a simple script that is similar to `kubectl create -f secret.yaml` but inserts into vault instead of apiserver.